engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/AmbiguousImportsAnalysis.scala

```diff
@@ -0,0 +1,430 @@
+package org.enso.compiler.pass.analyse
+
+import org.enso.interpreter.runtime.Module
+import org.enso.compiler.context.{InlineContext, ModuleContext}
+import org.enso.compiler.core.IR
+import org.enso.compiler.core.IR.Module.Scope.Import
+import org.enso.compiler.core.IR.Module.Scope.Import.Polyglot
+import org.enso.compiler.data.BindingsMap
+import org.enso.compiler.exception.CompilerError
+import org.enso.compiler.pass.IRPass
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+/** A pass that checks for ambiguous and duplicated symbols from imports.
+  * A duplicated import is an import of a symbol that has already been imported and refers to the
+  * same object (entity). On the other hand, an ambiguous import is an import of a symbol that has already
+  * been imported but refers to a different object. For every duplicated import, a warning is attached
+  * to the IR, and for every ambiguous import, the IR is replaced with an error.
+  * To identify an object, this pass uses physical path of the object instead of the object itself.
+  *
+  * One import IR can be replaced with multiple error IRs. This is the case for `from ... import ...`
+  * import statements.
+  *
+  * The original import is saved in the error and warning so that the user can see from which location
+  * the symbol was originally imported.
+  *
+  * Also iterates polyglot imports.
+  *
+  * All synthetic imports and exports, as well as synthetic modules are ignored by this pass.
+  *
+  * Thi pass does not alter any metadata.
+  */
+case object AmbiguousImportsAnalysis extends IRPass {
+
+  override type Metadata = IRPass.Metadata.Empty
+
+  override type Config = IRPass.Configuration.Default
+
+  override val precursorPasses: Seq[IRPass] =
+    Seq(BindingAnalysis, ImportSymbolAnalysis)
+
+  override val invalidatedPasses: Seq[IRPass] =
+    Seq()
+
+  /** @inheritdoc
+    */
+  override def runExpression(
+    ir: IR.Expression,
+    inlineContext: InlineContext
+  ): IR.Expression = ir
+
+  /** @inheritdoc
+    */
+  override def updateMetadataInDuplicate[T <: IR](sourceIr: T, copyOfIr: T): T =
+    copyOfIr
+
+  /** @inheritdoc
+    */
+  override def runModule(
+    ir: IR.Module,
+    moduleContext: ModuleContext
+  ): IR.Module = {
+    val module = moduleContext.module
+    if (module.isSynthetic) {
+      ir
```

Comment:
What's the justification of ignoring _synthetic_ modules? Are they known to be correct with respect to duplicates and ambiguities? Or are they skipped because they are known to contain duplicates?

Reply:
*synthetic* modules contain only *synthetic* imports and exports that were mostly generated in `Compiler.injectSyntheticModuleExports`. There are duplications, but I believe these duplications should not be presented to the end user. Moreover, there are no ambiguous imports there - I checked that - and also, we have that fully under control.